### PR TITLE
Fix/distribute header

### PR DESCRIPTION
**Pull request by ameenhaq**
* Set header in run function to be distributed across executors - as part of Context Object.

### CODE REVIEW

File: repository-hbase/src/main/java/hbase/load/BulkLoader.java [hunk 1]
```diff
@@ -97,11 +107,40 @@ public int run(String[] strings) throws Exception {
...
 
-    private int load(String tableNameStr, String referencePeriod, String inputFile, String outputFilePath) {
+    private void getHeader(String inputFile, Configuration conf) throws Exception {
+        Path path = new Path(inputFile);
+        FileSystem fs = FileSystem.get(path.toUri(), conf);
+
+        try( BufferedReader readFile = new BufferedReader(new InputStreamReader(fs.open(path))) ) {
+            LOG.debug("Successfully read file at '{}' and now getting header.", inputFile);
+            String header;
+            try {
```
Comment (awharris) [inline repository-hbase/src/main/java/hbase/load/BulkLoader.java]: Nested try not necessary, will be caught by outer try

File: repository-hbase/src/main/java/hbase/load/BulkLoader.java [hunk 2]
```diff
@@ -97,11 +107,40 @@ public int run(String[] strings) throws Exception {
...
+    private int load(String tableNameStr, String referencePeriod, String inputFile) throws Exception {
         return load(tableNameStr, referencePeriod, inputFile, "");
     }
 
-    private int load(String tableNameStr, String referencePeriod, String inputFile, String outputFilePath) {
+    private void getHeader(String inputFile, Configuration conf) throws Exception {
+        Path path = new Path(inputFile);
+        FileSystem fs = FileSystem.get(path.toUri(), conf);
+
+        try( BufferedReader readFile = new BufferedReader(new InputStreamReader(fs.open(path))) ) {
```
Comment (awharris) [inline repository-hbase/src/main/java/hbase/load/BulkLoader.java]: InputStreamReader not being closed?

Change to...

try( 
         InputStreamReader  reader = new InputStreamReader(fs.open(path));
 BufferedReader readFile = new BufferedReader(reader)
    )

File: repository-hbase/src/main/java/hbase/load/BulkLoader.java [hunk 3]
```diff
@@ -97,11 +107,40 @@ public int run(String[] strings) throws Exception {
...
+                    System.out.println("Found header " +  header);
+                    conf.set(COLUMN_HEADINGS, header);
+                }
+                else {
+                    System.out.println("Header is NUll - for reading " + inputFile);
+                    LOG.error("Header is NUll - for reading '{}'", inputFile);
+                    throw new Exception("Header is NUll - for reading " + inputFile);
+                }
+            }
+            catch (Exception ex) {
```
Comment (awharris) [inline repository-hbase/src/main/java/hbase/load/BulkLoader.java]: Remove this try/catch

File: repository-hbase/src/main/java/hbase/load/BulkLoader.java [hunk 4]
```diff
@@ -97,11 +107,40 @@ public int run(String[] strings) throws Exception {
...
+            String header;
+            try {
+                header = readFile.readLine();
+                if (header != null) {
+                    LOG.debug("Found header '{}'", header);
+                    System.out.println("Found header " +  header);
+                    conf.set(COLUMN_HEADINGS, header);
+                }
+                else {
+                    System.out.println("Header is NUll - for reading " + inputFile);
```
Comment (awharris) [inline repository-hbase/src/main/java/hbase/load/BulkLoader.java]: Remove system.out, only use logging

File: repository-hbase/src/main/java/hbase/load/CSVDataKVMapper.java [hunk 5]
```diff
@@ -151,7 +153,7 @@ private void writeRow(Context context, String rowKeyStr, String[] fields) throws
                 try {
                     put.add(new KeyValue(rowKey.get(), columnFamily, columnNames[i], fields[i].getBytes()));
                 } catch (Exception e) {
-                    LOG.error("Cannot add line '{}'", fields[0], e);
+                    LOG.error("ColumnName: '{}' \n.Cannot add line '{}' with header '{}'", columnNames, fields[0], columnNames[i], e);
```
Comment (awharris) [inline repository-hbase/src/main/java/hbase/load/CSVDataKVMapper.java]: Believe errors need to be added to a central counter in order to report back at the end when results are collated post distribution. Ask Ranjit

File: repository-hbase/src/main/java/hbase/load/CSVDataKVMapper.java [hunk 6]
```diff
@@ -151,7 +153,7 @@ private void writeRow(Context context, String rowKeyStr, String[] fields) throws
...
                     throw e;
```
Comment (awharris) [inline repository-hbase/src/main/java/hbase/load/CSVDataKVMapper.java]: Don't think we want to fail the whole process if one record fails to write so don't throw an exception just log the error and move on unless the error is fatal

**Review by awharris — CHANGES_REQUESTED** [review]

File: repository-hbase/src/main/java/hbase/load/CSVDataKVMapper.java (hunk 5 above)
Comment (ranjit-das) [reply repository-hbase/src/main/java/hbase/load/CSVDataKVMapper.java]: I have noticed the BAD_HBASE_RECORDS counter is incremented if there is any failure while writing to HFile. So I think its ok if we don't increase the counter here while processing all the fields of a record.

**Review by awharris — APPROVED** [review]

**Pull request merged**